pkg/venafi/firefly/deviceFlow.go

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2023 Venafi, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package firefly
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/Venafi/vcert/v5/pkg/endpoint"
+	"github.com/Venafi/vcert/v5/pkg/verror"
+	"golang.org/x/oauth2"
+)
+
+// DeviceCred It's the representation of the info returned when a Device Code is requested
+// to the OAuth 2.0 Identity Provider to request an access code
+type DeviceCred struct {
+	DeviceCode      string `json:"device_code"`
+	UserCode        string `json:"user_code"`
+	VerificationURL string `json:"verification_url"` //Google use this to return the URL to share to the user
+	VerificationURI string `json:"verification_uri"` // others like Okta, Auth0 and WSO2 use this one to return the URI to share to the user
+	Interval        int64  `json:"interval"`
+	ExpiresIn       int64  `json:"expires_in"`
+}
+
+func (c *Connector) getDeviceAccessToken(auth *endpoint.Authentication) (*oauth2.Token, error) {
+
+	//requesting the device code
+	devCred, err := c.requestDeviceCode(auth)
+
+	if err != nil {
+		return nil, err
+	}
+
+	//setting as default verificationURL the value returned in VerificationURI given that is the used for the most of the
+	// OAuth IdP like Okta, Auth0 and WSO2 return the verification_uri to show to the user
+	verificationURL := devCred.VerificationURI
+	//if that is empty then trying to use the VerificationURL given google uses verification_url to
+	//return the verification_uri to show to the user
+	if verificationURL == "" {
+		verificationURL = devCred.VerificationURL
+	}
+
+	fmt.Printf("please go to %v and enter the code %v . It will expire in %dm and %ds\n\n", verificationURL, devCred.UserCode, devCred.ExpiresIn/60, devCred.ExpiresIn%60)
```

Comment:
We should make a bigger and better message here as this is really important the user see it.
One example is how aws-cli does it.

Reply:
done